src/bennu/devices/modules/comms/iec60870-5/module/Server.cpp

```diff
@@ -174,6 +233,69 @@ void Server::reversePoll()
     }
 }
 
+/*
+ * Reverse poll loop that sends local bennu datastore data to
+ * connected clients. Sends indications as double point values
+ */
+void Server::reversePollDoublePoint()
+{
+    while (1)
+    {
+        if (mConnected)
+        {
+            CS101_AppLayerParameters alParams = IMasterConnection_getApplicationLayerParameters(mConnection);
+
+            // Send binary values
+            // kv = {<address>: {<tag>, eInput}}
+            for (const auto &kv : gServer->mBinaryPoints)
+            {
+                const std::string tag = kv.second.first;
+                sendSpontaneousUpdate(mConnection, kv.first, convertBoolToDPValue(gServer->mDataManager->getDataByTag<bool>(tag)));
+            }
+
+            // Send analog values
+            CS101_ASDU newAsdu = CS101_ASDU_create(alParams, false, CS101_COT_PERIODIC, 0, 1, false, false);
+            // kv = {<address>: {<tag>, eInput}}
+            for (const auto &kv : gServer->mAnalogPoints)
+            {
+                const std::string tag = kv.second.first;
+                if (CS101_ASDU_getPayloadSize(newAsdu) < MAX_ASDU_PAYLOAD_SIZE)
```

Comment:
As far as I can tell, the inner if statement `if (gServer->mDataManager->hasTag(tag))` is duplicated. Could this be inverted to handle just the `else`, then have the duplicated logic afterward?

Possibly something like this:

```c++
if (CS101_ASDU_getPayloadSize(newAsdu) >= MAX_ASDU_PAYLOAD_SIZE)
{
                    IMasterConnection_sendASDU(mConnection, newAsdu);
                    CS101_ASDU_destroy(newAsdu);
                    newAsdu = CS101_ASDU_create(alParams, false, CS101_COT_PERIODIC, 0, 1, false, false);
                    if (gServer->mDataManager->hasTag(tag))
}

if (gServer->mDataManager->hasTag(tag))
{
    auto val = gServer->mDataManager->getDataByTag<double>(tag);
    InformationObject io = (InformationObject)MeasuredValueShort_create(NULL, kv.first, val, IEC60870_QUALITY_GOOD);
    CS101_ASDU_addInformationObject(newAsdu, io);
    InformationObject_destroy(io);
}
```

Reply:
fixed